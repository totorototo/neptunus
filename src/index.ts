export {
  Position,
  Elevation,
  Longitude,
  Latitude,
  PathAnalyst,
  PositionAnalyst,
  Path,
  Area
} from './types';

export { calculateDistance } from './utils/helper';

export { createPositionAnalyst } from './createPositionAnalyst';

export { createPathAnalyst } from './createPathAnalyst';
