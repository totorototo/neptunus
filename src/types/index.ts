import { Area } from './area';
import { Position, Altitude, Longitude, Latitude } from './position';
import { Track } from './track';
import { Elevation } from './elevation';
import { Statistics, Distance, Gain, Loss } from './statistics';

export {
  Area,
  Position,
  Altitude,
  Longitude,
  Latitude,
  Track,
  Elevation,
  Statistics,
  Gain,
  Loss,
  Distance
};
