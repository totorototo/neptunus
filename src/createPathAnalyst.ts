import { Position } from './types/position';
import { PathAnalyst } from './types/pathAnalyst';
import { Path } from './types/path';
import { calculateDistance } from './utils/helper';

type Statistics = [number, number, number, number];
type Elevation = { positive: number; negative: number };
type Region = {
  minLongitude: number;
  maxLongitude: number;
  minLatitude: number;
  maxLatitude: number;
};
type Internals = {
  distance: number;
  gain: number;
  loss: number;
  map: Statistics[];
};

export const createPathAnalyst = (path: Path): PathAnalyst => {
  const mapLocationsToProgression = (): Internals =>
    path.reduce(
      (
        accu: {
          distance: number;
          gain: number;
          loss: number;
          map: [Statistics];
        },
        location: Position,
        index: number,
        array: Position[]
      ) => {
        if (index > 0) {
          const distance = calculateDistance(array[index - 1], location);
          accu.distance += distance;
          const destinationElevation = array[index - 1][2];
          const originElevation = location[2];
          if (originElevation && destinationElevation) {
            const Δφ = destinationElevation - destinationElevation;
            if (Δφ > 0) {
              accu.gain += Δφ;
            } else {
              accu.loss += Math.abs(Δφ);
            }
          }

          accu.map.push([index, accu.distance, accu.gain, accu.loss]);
          return accu;
        } else {
          return accu;
        }
      },
      { distance: 0, gain: 0, loss: 0, map: [[0, 0, 0, 0]] }
    );

  const calculatePathLength = (): number =>
    path.reduce(
      (distance, position, index, array) =>
        index < array.length - 1
          ? distance + calculateDistance(position, array[index + 1])
          : distance,
      0
    );

  const calculatePathElevation = (smoothingFactor = 5): Elevation => {
    const elevations = path.map(position => position[2] || 0);

    // smooth array values (remove noise).
    const smoothedElevations = elevations.reduce(
      (
        elevations: number[],
        elevation: number,
        index: number,
        values: number[]
      ) => {
        if (
          values[index - smoothingFactor] &&
          values[index + smoothingFactor]
        ) {
          const filteredElevations = values.slice(
            index - smoothingFactor,
            index + smoothingFactor
          );
          const elevationGain = filteredElevations.reduce(
            (sum, item) => sum + item,
            0
          );

          const averageElevationGain = elevationGain / (smoothingFactor * 2);
          elevations.push(averageElevationGain);
          return elevations;
        }
        elevations.push(elevation);
        return elevations;
      },
      []
    );
    return smoothedElevations.reduce(
      (elevationGain, elevation, index, values) => {
        if (values[index + 1]) {
          const Δφ = values[index + 1] - elevation;
          if (Δφ > 0) {
            elevationGain.positive = elevationGain.positive + Δφ;
          } else {
            elevationGain.negative = elevationGain.negative + Math.abs(Δφ);
          }
          return elevationGain;
        }
        return elevationGain;
      },
      { positive: 0, negative: 0 }
    );
  };

  const calculatePathBoundingBox = (): Region =>
    path.reduce(
      (region, position) => ({
        minLongitude:
          position[0] < region.minLongitude ? position[0] : region.minLongitude,
        maxLongitude:
          position[0] > region.maxLongitude ? position[0] : region.maxLongitude,
        minLatitude:
          position[1] < region.minLatitude ? position[1] : region.minLatitude,
        maxLatitude:
          position[1] > region.maxLatitude ? position[1] : region.maxLatitude
      }),
      {
        minLongitude: path[0][0],
        maxLongitude: path[0][0],
        minLatitude: path[0][1],
        maxLatitude: path[0][1]
      }
    );

  const findClosestPosition = (currentPosition: Position): Position => {
    const closestLocation = path.reduce(
      (accu, position) => {
        const distance = calculateDistance(position, currentPosition);

        if (distance < accu.distance) {
          accu.distance = distance;
          accu.position = position;
        }
        return accu;
      },
      {
        position: path[0],
        distance: calculateDistance(currentPosition, path[0])
      }
    );
    return closestLocation.position;
  };

  const getPositionIndex = (position: Position): number =>
    path.findIndex(
      currentPosition =>
        currentPosition[0] === position[0] &&
        currentPosition[1] === position[1] &&
        currentPosition[2] === position[2]
    );

  const findClosestIndex = (map: Statistics[], distance: number): number => {
    return map.reduce((accu, item) => {
      if (accu === null) return item;
      if (Math.abs(distance - item[1]) >= Math.abs(distance - accu[1])) {
        return accu;
      }
      return item;
    })[0];
  };

  const getPositionsIndicesAlongPath = (...distances: number[]): number[] => {
    const mapIndexProgression = mapLocationsToProgression();
    return distances.map(distance => {
      return findClosestIndex(mapIndexProgression.map, distance);
    });
  };

  const getPositionsAlongPath = (...distances: number[]): Position[] =>
    getPositionsIndicesAlongPath(...distances).map(index => path[index]);

  const splitPath = (start = 0, end = 0): Position[] => {
    const locationsIndices = getPositionsIndicesAlongPath(start, end);
    const splitTrace = path.slice(locationsIndices[0], locationsIndices[1]);

    return splitTrace;
  };

  return {
    getPositionsAlongPath,
    getPositionsIndicesAlongPath,
    getPositionIndex,
    splitPath,
    calculatePathLength,
    calculatePathElevation,
    calculatePathBoundingBox,
    findClosestPosition
  };
};
