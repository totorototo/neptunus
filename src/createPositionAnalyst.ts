import { PositionAnalyst, Position, Area } from './types';
import { calculateDistance } from './utils/helper';

export const createPositionAnalyst = (position: Position): PositionAnalyst => {
  const isInArea = (area: Area): boolean => {
    return (
      position[0] > area.minLongitude &&
      position[0] < area.maxLongitude &&
      position[1] > area.minLatitude &&
      position[1] < area.maxLatitude
    );
  };

  const isInRadius = (center: Position, radius: number): boolean => {
    const distance = calculateDistance(center, position);
    return distance - radius < 0;
  };
  const distanceFromPosition = (destination: Position): number => {
    return calculateDistance(position, destination);
  };
  return { isInArea, isInRadius, distanceFromPosition };
};
