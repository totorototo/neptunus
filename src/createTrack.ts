import { Area, Elevation, Track, Position, Statistics } from './types';
import { calculateDistance } from './utils/helper';

type Internals = {
  distance: number;
  gain: number;
  loss: number;
  map: Statistics[];
};

export const createTrack = (positions: Position[]): Track => {
  const mapPositionsToProgression = (): Internals =>
    positions.reduce(
      (
        accum: {
          distance: number;
          gain: number;
          loss: number;
          map: [Statistics];
        },
        location: Position,
        index: number,
        array: Position[]
      ) => {
        if (index > 0) {
          const distance = calculateDistance(array[index - 1], location);
          accum.distance += distance;
          const destinationElevation = location[2];
          const originElevation = array[index - 1][2];
          if (originElevation && destinationElevation) {
            const delta = destinationElevation - originElevation;
            if (delta > 0) {
              accum.gain += delta;
            } else {
              accum.loss += Math.abs(delta);
            }
          }
          accum.map.push([accum.distance, accum.gain, accum.loss]);
          return accum;
        } else {
          return accum;
        }
      },
      { distance: 0, gain: 0, loss: 0, map: [[0, 0, 0]] }
    );

  const getLength = (): number =>
    positions.reduce(
      (distance, position, index, array) =>
        index < array.length - 1
          ? distance + calculateDistance(position, array[index + 1])
          : distance,
      0
    );

  const getElevation = (smoothingFactor = 5): Elevation => {
    const elevations = positions.map(position => position[2] || 0);

    // smooth array values (remove noise).
    const smoothedElevations = elevations.reduce(
      (
        elevations: number[],
        elevation: number,
        index: number,
        values: number[]
      ) => {
        if (
          values[index - smoothingFactor] &&
          values[index + smoothingFactor]
        ) {
          const filteredElevations = values.slice(
            index - smoothingFactor,
            index + smoothingFactor
          );
          const elevationGain = filteredElevations.reduce(
            (sum, item) => sum + item,
            0
          );

          const averageElevationGain = elevationGain / (smoothingFactor * 2);
          elevations.push(averageElevationGain);
          return elevations;
        }
        elevations.push(elevation);
        return elevations;
      },
      []
    );
    return smoothedElevations.reduce(
      (elevationGain, elevation, index, values) => {
        if (values[index + 1]) {
          const delta = values[index + 1] - elevation;
          if (delta > 0) {
            elevationGain.positive = elevationGain.positive + delta;
          } else {
            elevationGain.negative = elevationGain.negative + Math.abs(delta);
          }
          return elevationGain;
        }
        return elevationGain;
      },
      { positive: 0, negative: 0 }
    );
  };

  const getBoundingBox = (): Area =>
    positions.reduce(
      (region, position) => ({
        minLongitude: Math.min(position[0], region.minLongitude),
        maxLongitude: Math.max(position[0], region.maxLongitude),
        minLatitude: Math.min(position[1], region.minLatitude),
        maxLatitude: Math.max(position[1], region.maxLatitude)
      }),
      {
        minLongitude: positions[0][0],
        maxLongitude: positions[0][0],
        minLatitude: positions[0][1],
        maxLatitude: positions[0][1]
      }
    );

  const findClosestPosition = (currentPosition: Position): Position => {
    const closestLocation = positions.reduce(
      (accum, position) => {
        const distance = calculateDistance(position, currentPosition);

        if (distance < accum.distance) {
          accum.distance = distance;
          accum.position = position;
        }
        return accum;
      },
      {
        position: positions[0],
        distance: calculateDistance(currentPosition, positions[0])
      }
    );
    return closestLocation.position;
  };

  const getPositionIndex = (position: Position): number =>
    positions.findIndex(
      currentPosition =>
        currentPosition[0] === position[0] &&
        currentPosition[1] === position[1] &&
        currentPosition[2] === position[2]
    );

  const findClosestIndex = (map: Statistics[], distance: number): number => {
    const { index } = map.reduce(
      (accum, item, index) => {
        if (
          Math.abs(distance - item[0]) >= Math.abs(distance - accum.stat[0])
        ) {
          return accum;
        }
        return { ...accum, index: index, stat: item };
      },
      { index: 0, stat: [0, 0, 0] }
    );

    return index;
  };

  const getPositionsIndicesAt = (...distances: number[]): number[] => {
    const mapIndexProgression = mapPositionsToProgression();
    return distances.map(distance => {
      return findClosestIndex(mapIndexProgression.map, distance);
    });
  };

  const getPositionsAt = (...distances: number[]): Position[] =>
    getPositionsIndicesAt(...distances).map(index => positions[index]);

  const slice = (start = 0, end = 0): Position[] => {
    const locationsIndices = getPositionsIndicesAt(start, end);
    return positions.slice(locationsIndices[0], locationsIndices[1]);
  };

  const getProgressionStatistics = (currentPathIndex: number): Statistics => {
    const mapIndexProgression = mapPositionsToProgression();
    const result = mapIndexProgression.map.find(
      (_, index) => index === currentPathIndex
    );
    return result ? result : [0, 0, 0];
  };

  return {
    getProgressionStatistics,
    getPositionsAt,
    getPositionsIndicesAt,
    getPositionIndex,
    slice,
    getLength,
    getElevation,
    getBoundingBox,
    findClosestPosition
  };
};
