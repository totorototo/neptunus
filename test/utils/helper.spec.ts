import { calculateDistance, calculateBearing, Position } from '../../src/index';

describe('helper', () => {
  it('calcute distance between Paris and Moscow', () => {
    const origin: Position = [37.618423, 55.751244];
    const destination: Position = [2.3488, 48.8534];

    const distance = calculateDistance(origin, destination);
    expect(distance).toMatchSnapshot();
  });

  it('calculate bearing between two positions', () => {
    const origin: Position = [-3.94915, 51.2194];
    const destination: Position = [-3.95935, 51.2392];

    const bearing = calculateBearing(origin, destination);
    expect(bearing).toMatchSnapshot();
  });
});
