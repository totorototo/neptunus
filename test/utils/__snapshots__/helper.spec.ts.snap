// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper calculate bearing between two positions 1`] = `58.65519236183434`;

exports[`helper calculate distance between Paris and Moscow 1`] = `2486688.9750256147`;
